### ISSUE TITLE
Repeatedly push an view controller whose view owned a apng animation view and pop it, app will crash.

### ISSUE THREAD

**Issue by Desgard** [onevcat/APNGKit#56]
This is the occasional problem. And we detected it by our apm system. And finally positioned to the **APNGKit**. I test the APNGKit's demo, it also has the same problem.

This is the error symbol stack log of *APNGKitDemo* proj.

```console
Hardware Model:      iPhone7,2
Process:             APNGDemo-iOS [3710]
Path:                /private/var/containers/Bundle/Application/42CDB407-533D-4EA7-8F5D-77F199CBAC8B/APNGDemo-iOS.app/APNGDemo-iOS
Identifier:          com.onevcat.APNGDemo-iOS
Version:             1 (1.0)
Code Type:           ARM-64 (Native)
Role:                Foreground
Parent Process:      launchd [1]
Coalition:           com.onevcat.APNGDemo-iOS [3835]


Date/Time:           2017-11-07 10:28:59.6368 +0800
Launch Time:         2017-11-07 10:27:52.6125 +0800
OS Version:          iPhone OS 10.3.2 (14F89)
Report Version:      104

Exception Type:  EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0x0000000109e38000
Termination Signal: Segmentation fault: 11
Termination Reason: Namespace SIGNAL, Code 0xb
Terminating Process: exc handler [0]
Triggered by Thread:  0

Filtered syslog:
None found

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0 Crashed:
0   libsystem_platform.dylib      	0x000000018bd20e10 _platform_memmove + 176
1   CoreGraphics                  	0x000000018e0702c0 decode_data + 13144
2   CoreGraphics                  	0x000000018e239720 img_decode_read + 2024
3   CoreGraphics                  	0x000000018e23d490 img_alphamerge_read + 548
4   CoreGraphics                  	0x000000018e2408fc img_data_lock + 6124
5   CoreGraphics                  	0x000000018e23f0b8 CGSImageDataLock + 176
6   CoreGraphics                  	0x000000018e05f1c0 ripc_AcquireImage + 756
7   CoreGraphics                  	0x000000018e2539a4 ripc_DrawImage + 656
8   CoreGraphics                  	0x000000018e243b50 CGContextDrawImageWithOptions + 632
9   QuartzCore                    	0x000000018fe87a8c CA::Render::(anonymous namespace)::create_image_by_rendering+ 105100 (CGImage*, CGColorSpace*, unsigned int, double) + 1068
10  QuartzCore                    	0x000000018fe88960 CA::Render::(anonymous namespace)::create_image_from_rgb_image+ 108896 (CGImage*, CGColorSpace*, unsigned int, double) + 660
11  QuartzCore                    	0x000000018fe874b0 CA::Render::create_image+ 103600 (CGImage*, CGColorSpace*, unsigned int, double) + 876
12  QuartzCore                    	0x000000018fe89238 CA::Render::copy_image+ 111160 (CGImage*, CGColorSpace*, unsigned int, double, double) + 472
13  QuartzCore                    	0x000000018ff87434 -[CALayer+ 1152052 (CALayerPrivate) _copyRenderLayer:layerFlags:commitFlags:] + 480
14  QuartzCore                    	0x000000018fef4a74 CA::Context::commit_layer+ 551540 (CA::Layer*, unsigned int, unsigned int, void*) + 108
15  QuartzCore                    	0x000000018ff7a3a8 CA::Layer::commit_if_needed(CA::Transaction*, void (*)+ 1098664 (CA::Layer*, unsigned int, unsigned int, void*), void*) + 388
16  QuartzCore                    	0x000000018feb39ac x_hash_table_foreach + 72
17  QuartzCore                    	0x000000018ff1d4c0 CA::Transaction::foreach_root(void (*)+ 718016 (CA::Layer*, void*), void*) + 40
18  QuartzCore                    	0x000000018fef5778 CA::Context::commit_transaction+ 554872 (CA::Transaction*) + 1320
19  QuartzCore                    	0x000000018ff1c3ac CA::Transaction::commit+ 713644 () + 504
20  QuartzCore                    	0x000000018ff1ce78 CA::Transaction::observer_callback+ 716408 (__CFRunLoopObserver*, unsigned long, void*) + 120
21  CoreFoundation                	0x000000018cc149a8 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 32
22  CoreFoundation                	0x000000018cc12630 __CFRunLoopDoObservers + 372
23  CoreFoundation                	0x000000018cc12a7c __CFRunLoopRun + 956
24  CoreFoundation                	0x000000018cb42da4 CFRunLoopRunSpecific + 424
25  GraphicsServices              	0x000000018e5ac074 GSEventRunModal + 100
26  UIKit                         	0x0000000192dfd058 UIApplicationMain + 208
27  APNGDemo-iOS                  	0x0000000100097cfc 0x100090000 + 31996
28  libdyld.dylib                 	0x000000018bb5159c start + 4
```

**Comment by Desgard** [comment]
This is the export log file.

[APNGDemo-iOS  07-11-2017, 10-28.crash.zip](https://github.com/onevcat/APNGKit/files/1448513/APNGDemo-iOS.07-11-2017.10-28.crash.zip)

**Comment by onevcat** [comment]
This issue exists deeply in the CoreGraphics. I didn't see any problem in APNGKit for now, so it seems there is little we could do it yet. But any further information on this is welcome and it might help to dig into the problem.

I tried on both iOS device and simulator, and reproduce it once in a real device (after several hundreds of pushing and popping). However, I can never reproduce this problem on a simulator. @Desgard is it also your case?

**Comment by 1jacobb1** [comment]
I also encountered this kind of crash regarding the CoreGraphics

**Comment by Cierpliwy** [comment]
We had crashes with _platform_memmove function present in call stack. It was due to usage of GCDTimer and happened only after deallocation of APNGImageView. We migrated to CADisplayLink in our fork and didn't see any issues so far: https://github.com/Polidea/APNGKit.

**Comment by onevcat** [comment]
APNGKit v2 was released with fixing of dropping GCD timer, but using back to display link.

The new version also contains other countless fixes with a rewritten from scratch. If anyone sees a similar issue, please feel free to open a new issue!

Thanks.
